presto-main/src/main/java/com/facebook/presto/operator/scalar/MapConcatFunction.java

```diff
@@ -13,53 +13,167 @@
-            @TypeParameter("K") Type keyType,
-            @TypeParameter("V") Type valueType,
-            @SqlType("map(K,V)") Block leftMap,
-            @SqlType("map(K,V)") Block rightMap)
+    public static final MapConcatFunction MAP_CONCAT_FUNCTION = new MapConcatFunction();
+
+    private static final String FUNCTION_NAME = "map_concat";
+    private static final String DESCRIPTION = "Concatenates given maps";
+
+    private static final MethodHandle CONSTRUCTOR_METHOD_HANDLE = methodHandle(MapConcatFunction.class, "createMapState", Type.class, Type.class);
+    private static final MethodHandle METHOD_HANDLE = methodHandle(MapConcatFunction.class, "mapConcat", Type.class, Type.class, Object.class, Block[].class);
+
+    private MapConcatFunction()
+    {
+        super(new Signature(FUNCTION_NAME,
+                FunctionKind.SCALAR,
+                ImmutableList.of(typeVariable("K"), typeVariable("V")),
+                ImmutableList.of(),
+                parseTypeSignature("map(K,V)"),
+                ImmutableList.of(parseTypeSignature("map(K,V)")),
+                true));
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return true;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return DESCRIPTION;
+    }
+
+    @Override
+    public ScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
     {
-        if (leftMap.getPositionCount() == 0) {
-            return rightMap;
+        if (arity < 2) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "There must be two or more concatenation arguments");
         }
-        if (rightMap.getPositionCount() == 0) {
-            return leftMap;
+
+        Type keyType = boundVariables.getTypeVariable("K");
+        Type valueType = boundVariables.getTypeVariable("V");
+
+        MethodHandleAndConstructor methodHandleAndConstructor = generateVarArgsToArrayAdapter(
+                Block.class,
+                Block.class,
+                arity,
+                METHOD_HANDLE.bindTo(keyType).bindTo(valueType),
+                Optional.of(CONSTRUCTOR_METHOD_HANDLE.bindTo(keyType).bindTo(valueType)));
+
+        return new ScalarFunctionImplementation(
+                false,
+                nCopies(arity, false),
+                nCopies(arity, false),
+                methodHandleAndConstructor.getMethodHandle(),
+                Optional.of(methodHandleAndConstructor.getConstructor()),
+                isDeterministic());
+    }
+
+    @UsedByGeneratedCode
+    public static Object createMapState(Type keyType, Type valueType)
+    {
+        return new PageBuilder(ImmutableList.of(keyType, valueType));
+    }
+
+    @UsedByGeneratedCode
+    public static Block mapConcat(Type keyType, Type valueType, Object state, Block[] maps)
```

Comment:
I think `TypedSet` should not be put into state since it is used to deduplicate the keys and it should be different for different rows. And there is no method like `clear` to remove all elements in `TypedSet`.